blog/2024-01-25-inspecting-dependency-inference-results/index.mdx

```diff
@@ -0,0 +1,152 @@
+---
+    authors: [alexey]
+    tags: [concepts]
+---
+
+# Inspecting dependency inference results
+
+<CaptionedImg src={require("./splash.jpg").default}>
+  Photo by [Agence
+  Olloweb](https://unsplash.com/photos/magnifying-glass-near-gray-laptop-computer-d9ILr-dbEdg)
+  /
+  [Unsplash](https://unsplash.com/photos/magnifying-glass-near-gray-laptop-computer-d9ILr-dbEdg)
+</CaptionedImg>
+
+{/* truncate */}
+
+An important distinguishing feature of Pants v2 is its ability to automatically [infer your code's internal and external dependencies](https://www.pantsbuild.org/blog/2022/10/27/why-dependency-inference). The information about them is available via the `dependencies` and `peek` goals, so that metadata about what modules import from modules is always readily accessible.
+
+While managing a codebase in a language supported by Pants such as Python, Java, or Scala, it can, however, be useful to go further and explore your import statements to compare them with what has been inferred by Pants. This may be particularly helpful for large modules with lots of imports where it's easy to miss the fact that a dependency is omitted or that there's a dependency inferred that should really be ignored. To get help with that, you can use one of the less known goals, [`python-dump-source-analysis`](https://www.pantsbuild.org/2.18/reference/goals/python-dump-source-analysis), which can dump source code analysis for the Python source code build targets.
```

Comment:
```suggestion
While managing a codebase in a [Pants-supported language](https://www.pantsbuild.org/2.18/docs/introduction/welcome-to-pants#which-languages-and-frameworks-does-pants-support) (e.g. Python, Java, Scala, etc...), it can sometimes be useful dig deeper and explore the inferred dependencies in order to compare them to your source code's import statements. This can be particularly useful for large modules with many imports, where it's very easy to miss an omitted dependency or an inferred dependency that should have been ignored. 

You can start your exploration with one of the lesser known goals, [`python-dump-source-analysis`](https://www.pantsbuild.org/2.18/reference/goals/python-dump-source-analysis). This goal will dump the dependency analysis for the Python source code build targets.
```